lib/temporal/workflow/context.rb

```diff
@@ -215,14 +215,53 @@ def wait_for_all(*futures)
         return
       end
 
-      def wait_for(future)
+      # Block workflow progress until any future is finished or any unblock_condition
+      # block evaluates to true.
+      def wait_for(*futures, &unblock_condition)
+        if futures.empty? && unblock_condition.nil?
+          raise 'You must pass either a future or an unblock condition block to wait_for'
+        end
+
         fiber = Fiber.current
+        should_yield = false
+        blocked = true
+
+        if futures.any?
+          if futures.all?(&:finished?)
+            blocked = false
+          else
+            should_yield = true
+            futures.each do |future|
+              dispatcher.register_handler(future.target, Dispatcher::WILDCARD) do
+                if blocked && future.finished?
+                  # Because this block can run for any dispatch, ensure the fiber is only
+                  # resumed one time by checking if it's already been unblocked.
+                  blocked = false
+                  fiber.resume
+                end
+              end
+            end
+          end
+        end
 
-        dispatcher.register_handler(future.target, Dispatcher::WILDCARD) do
-          fiber.resume if future.finished?
+        if unblock_condition
+          if unblock_condition.call
+            blocked = false
+          else
+            should_yield = true
+
+            dispatcher.register_handler(Dispatcher::WILDCARD, Dispatcher::WILDCARD) do
```

Comment:
There's a slight problem here — we're not de-registering the handler afterwards, which means that it will be called for every event until the end of a workflow. It won't be doing anything because of that `blocked` variable, but something we might want to address. It wasn't an issue previously because each target was expected to fire each event only once, but still keeps these references in memory.

This can be done in a follow-up, I might have a look at it

Reply:
I suppose once it's unblocked, it can be unregistered. The interplay on this is messy because it needs a callback. I'll see if I can do this cleanly or if it's best to revise later.